cmd/restic/cmd_check.go

```diff
@@ -233,27 +264,44 @@ func runCheck(opts CheckOptions, gopts GlobalOptions, args []string) error {
 		}
 	}
 
-	doReadData := func(bucket, totalBuckets uint) {
-		packs := restic.IDSet{}
+	doSelectPacksByNumbers := func(bucket, totalBuckets uint) *restic.IDSet {
+		packs := &restic.IDSet{}
 		for pack := range chkr.GetPacks() {
 			// If we ever check more than the first byte
 			// of pack, update totalBucketsMax.
 			if (uint(pack[0]) % totalBuckets) == (bucket - 1) {
 				packs.Insert(pack)
 			}
 		}
-		packCount := uint64(len(packs))
+		return packs
+	}
 
-		if packCount < chkr.CountPacks() {
-			Verbosef(fmt.Sprintf("read group #%d of %d data packs (out of total %d packs in %d groups)\n", bucket, packCount, chkr.CountPacks(), totalBuckets))
-		} else {
-			Verbosef("read all data\n")
+	doSelectPacksByPercentage := func(percentage float64) *restic.IDSet {
+		packCount := len(chkr.GetPacks())
+		packsToCheck := int(math.Max((((float64(packCount)*percentage)/100.0)+0.5)-1.0, 0.0))
```

Comment:
Consider enforcing to check at least one pack if given percentage is too low but above 0.0.

Reply:
I think this should be either

```
packsToCheck := int(float64(packCount)*(percentage/100))
```

(current behavior), or

```
packsToCheck := int(math.Round(float64(packCount)*(percentage/100)))
```

I'm not sure why you're rounding down. The math.Max shouldn't do anything, if the rest of the code is correct. It can only obscure bugs elsewhere.

Reply:
Fixed in https://github.com/restic/restic/pull/3038/commits/4943b1ebda8ad41c5cd485e202336e780a1e8c77

I was thinking about array indices which is wrong in this case.

Reply:
But packsToCheck must be <= the number of packs, not <, because rand.Perm(n) generates integers < n. Without math.Round, if you ask for 99.99% of 2 packs, only 50% gets checked.